### PR TITLE
sklearn eval metric autologging: patch `metrics.SCORERS`

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes are proposed in this pull request?

sklearn eval metric autologging: patch `metrics.SCORERS`

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by dbczumar — COMMENTED** [review]
@WeichenXu123 I think I misinterpreted this PR before leaving a previous comment, so I'm editing it. Now, I'm wondering if we need the `finalizer` concept here or if we can perform `safe_patch()` on each value contained in `sklearn.metrics.SCORERS`. See https://github.com/mlflow/mlflow/pull/4638/files#r681224187

File: mlflow/utils/gorilla.py [hunk 1]
```diff
@@ -383,6 +391,9 @@ def revert(patch):
     if curr_active_patch in patch.destination.__dict__:
         delattr(patch.destination, curr_active_patch)
 
+    for finalizer in patch.finalizer_list:
+        finalizer()
```
Comment (dbczumar) [inline mlflow/utils/gorilla.py:395]: What happens if this throws?

File: mlflow/sklearn/__init__.py [hunk 2]
```diff
@@ -1494,10 +1496,24 @@ def should_patch_estimator_method(class_def, func_name):
...
+        scorer_fn_name = scorer_fn.__name__
+        try:
+            original_metric_fn = gorilla.get_original_attribute(sklearn.metrics, scorer_fn_name)
+        except AttributeError:
+            original_metric_fn = None
+        if scorer_fn is original_metric_fn:
+            scorer._score_func = getattr(sklearn.metrics, scorer_fn_name)
+            gorilla.get_active_patch(sklearn.metrics, scorer_fn_name).add_finalizer(
+                scorer_finalizer, scorer, original_metric_fn
+            )
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:1516]: Instead of adding finalizers, could we iterate through each of the scorers and `safe_patch` its `score_func` attribute?

```
for scorer in sklearn.metrics.SCORERS.values():
    safe_patch(FLAVOR_NAME, scorer, "_score_func", patched_metric_api, manage_run=False)
```

This seems to pass the tests you included in the PR.

Comment (WeichenXu123) [reply mlflow/sklearn/__init__.py:1516]: This is a good idea! Then no need adding hook when reverting patch.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !
